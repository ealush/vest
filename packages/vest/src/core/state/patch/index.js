import { getSuite } from '..';

/**
 * Updates current suite state with patcher value or output.
 * @param {string} suiteId.
 * @param {Function} patcher Uses to determine next value.
 * @return {Object} Next suite state.
 */
const patch = (suiteId, patcher) => {
  const [state, prevState] = getSuite(suiteId) ?? [];

  const nextState = patcher(state, prevState);

  if (nextState === state) {
    return state;
  }

  getSuite(suiteId)[0] = nextState;
  return nextState;
};

export default patch;
