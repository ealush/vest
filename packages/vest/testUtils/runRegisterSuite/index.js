import registerSuite from '../../src/core/state/registerSuite';
import runWithContext from '../../src/lib/runWithContext';

export default context => runWithContext(context, registerSuite);
