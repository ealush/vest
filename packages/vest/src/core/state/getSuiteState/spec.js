import { getState } from '..';
import runRegisterSuite from '../../../../testUtils/runRegisterSuite';
import { SYMBOL_SUITES } from '../symbols';
import getSuiteState from '.';

const suiteName = 'suite_1';
const suiteId = 'suiteId_1';
const suiteId_2 = 'suiteId_3';

describe('getSuiteState', () => {
  it('Should return current suite state by key', () => {
    runRegisterSuite({ name: suiteName, suiteId });
    runRegisterSuite({ name: suiteName, suiteId: suiteId_2 });
    const [currentState] = getState(SYMBOL_SUITES)[suiteId];
    expect(getSuiteState(suiteId)).toBe(currentState);
  });
});
