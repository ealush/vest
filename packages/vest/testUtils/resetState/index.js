import { OPERATION_MODE_STATEFUL } from '../../src/constants';
import state from '../../src/core/state';
import getSuiteState from '../../src/core/state/getSuiteState';
import { SYMBOL_STATE } from '../../src/core/state/symbols';
import singleton from '../../src/lib/singleton';
import runRegisterSuite from '../runRegisterSuite';

const resetState = (suiteId, suiteName = suiteId) => {
  singleton.set(SYMBOL_STATE, null);
  state.register();

  if (suiteName) {
    runRegisterSuite({
      suiteId,
      name: suiteName,
      operationMode: OPERATION_MODE_STATEFUL,
    });
    return getSuiteState(suiteId);
  }
};

export default resetState;
