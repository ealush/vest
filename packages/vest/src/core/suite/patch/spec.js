import * as state from '..';
import resetState from '../../../../testUtils/resetState';
import runRegisterSuite from '../../../../testUtils/runRegisterSuite';
import { OPERATION_MODE_STATEFUL } from '../../../constants';
import getSuiteState from '../getSuiteState';
import patch from '.';

const suiteName = 'suite_1';
const suiteId = 'suiteId_1';

let context;

describe('patch', () => {
  let patcher, suite, suiteState, prevState;

  beforeEach(() => {
    resetState();
    context = {
      name: suiteName,
      suiteId,
      operationMode: OPERATION_MODE_STATEFUL,
    };
  });

  describe('When patcher is a function', () => {
    beforeEach(() => {
      patcher = jest.fn(state => ({ ...state, ...{ k: 'v' } }));
      resetState();
      runRegisterSuite(context);
      suite = state.getSuite(suiteId);
      suiteState = suite[0];
      prevState = suite[1];
      patch(suiteId, patcher);
    });
    it('Should set current state value to patcher argument return value', () => {
      expect(state.getSuite(suiteId)[0]).toBe(patcher.mock.results[0].value);
    });
    it('Should pass current and prev state as the patcher arguments', () => {
      expect(patcher).toHaveBeenCalledWith(suiteState, prevState);
    });

    it('Should return next state', () => {
      expect(patch(suiteId, patcher)).toBe(getSuiteState(suiteId));
    });
  });
  describe('When patcher is not a function', () => {
    beforeEach(() => {
      patcher = {};
      resetState();
      runRegisterSuite(context);
    });

    it('Should throw an error', () => {
      expect(() => patch(suiteId, patcher)).toThrow(Error);
    });
  });
});
