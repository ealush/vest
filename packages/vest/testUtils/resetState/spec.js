import { getState } from '../../src/core/state';
import { SYMBOL_SUITES } from '../../src/core/state/symbols';
import resetState from '.';

describe('resetState', () => {
  test('Sanity', () => {
    expect(getState()).toMatchSnapshot();
  });

  it('Should create a new state', () => {
    const initialState = getState();
    expect(getState()).toBe(initialState);
    resetState();

    expect(getState()).not.toBe(initialState);
  });

  describe('When invoked with suite name', () => {
    let suiteId;

    beforeEach(() => {
      suiteId = 'test-suite';
    });

    it('Should add a new suite to the state', () => {
      expect(getState(SYMBOL_SUITES)).toEqual({});
      resetState(suiteId);
      expect(getState(SYMBOL_SUITES)[suiteId]).toBeTruthy();
      expect(getState(SYMBOL_SUITES)[suiteId]).toMatchSnapshot();
    });
  });
});
