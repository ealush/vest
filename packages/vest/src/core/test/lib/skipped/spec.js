import _ from 'lodash';
import resetState from '../../../../../testUtils/resetState';
import runRegisterSuite from '../../../../../testUtils/runRegisterSuite';
import getSuiteState from '../../../state/getSuiteState';
import patch from '../../../state/patch';
import { setSkipped, mergeSkipped } from '.';

const SUITE_NAME = 'suite_1';
const suiteId = 'suiteId_1';
const FIELD_NAME_1 = 'field_1';
const FIELD_NAME_2 = 'field_2';

const PREV_STATE_MOCK = {
  tests: {
    [FIELD_NAME_1]: {
      errorCount: 0,
      warnCount: 1,
      errors: [],
      warnings: ['warning_string'],
    },
    [FIELD_NAME_2]: {
      errorCount: 1,
      warnCount: 0,
      errors: ['error_string'],
      warnings: [],
    },
  },
};

describe('module: skipped', () => {
  beforeEach(() => {
    resetState(suiteId, SUITE_NAME);
  });

  describe('export: setSkipped', () => {
    describe('When field already skipped', () => {
      beforeEach(() => {
        setSkipped(suiteId, FIELD_NAME_1);
      });
      it('Should keep state unchanged', () => {
        const state = _.cloneDeep(getSuiteState(suiteId));
        setSkipped(suiteId, FIELD_NAME_1);
        expect(getSuiteState(suiteId)).toEqual(state);
      });
    });

    describe('When field is not yet skipped', () => {
      it('Should marked field as skipped', () => {
        expect(getSuiteState(suiteId).skipped[FIELD_NAME_1]).toBeUndefined();
        setSkipped(suiteId, FIELD_NAME_1);
        expect(getSuiteState(suiteId).skipped[FIELD_NAME_1]).toBe(true);
        expect(getSuiteState(suiteId)).toMatchSnapshot();
      });
    });
  });

  describe('export: mergeSkipped', () => {
    describe('When no previous state', () => {
      it('Should keep state unchanged', () => {
        const state = _.cloneDeep(getSuiteState(suiteId));
        mergeSkipped(suiteId);
        expect(getSuiteState(suiteId)).toEqual(state);
      });
    });

    describe('When previous state exists', () => {
      beforeEach(() => {
        runRegisterSuite({ name: SUITE_NAME, suiteId });
      });

      describe('When no currently skipped fields', () => {
        it('Should keep state unchanged', () => {
          const state = _.cloneDeep(getSuiteState(suiteId));
          mergeSkipped(suiteId);
          expect(getSuiteState(suiteId)).toEqual(state);
        });
      });

      describe('When skipped fields exist', () => {
        describe('When skipped field does not exist in previous state', () => {
          it('Should keep state unchanged', () => {
            const state = _.cloneDeep(getSuiteState(suiteId));
            mergeSkipped(suiteId);
            expect(getSuiteState(suiteId)).toEqual(state);
          });
        });

        describe('When skipped field exists in previous state', () => {
          let mock, state;
          beforeEach(() => {
            mock = _.cloneDeep(PREV_STATE_MOCK);
            patch(suiteId, state => _.merge({ ...state }, mock));
            runRegisterSuite({ name: SUITE_NAME, suiteId });
            state = _.cloneDeep(getSuiteState(suiteId));
          });

          describe('Warning field', () => {
            it('Should copy prev field state over', () => {
              expect(
                getSuiteState(suiteId).tests[FIELD_NAME_1]
              ).toBeUndefined();
              setSkipped(suiteId, FIELD_NAME_1);
              mergeSkipped(suiteId);
              expect(getSuiteState(suiteId).tests[FIELD_NAME_1]).toEqual(
                mock.tests[FIELD_NAME_1]
              );
              expect(getSuiteState(suiteId).warnCount).toBe(
                state.warnCount + mock.tests[FIELD_NAME_1].warnCount
              );
              expect(getSuiteState(suiteId).errorCount).toBe(state.errorCount);
              expect(getSuiteState(suiteId)).toMatchSnapshot();
            });
          });
          describe('Error field', () => {
            it('Should copy prev field state over', () => {
              expect(
                getSuiteState(suiteId).tests[FIELD_NAME_2]
              ).toBeUndefined();
              setSkipped(suiteId, FIELD_NAME_2);
              mergeSkipped(suiteId);
              expect(getSuiteState(suiteId).tests[FIELD_NAME_2]).toEqual(
                mock.tests[FIELD_NAME_2]
              );
              expect(getSuiteState(suiteId).errorCount).toBe(
                state.errorCount + mock.tests[FIELD_NAME_2].errorCount
              );
              expect(getSuiteState(suiteId).warnCount).toBe(state.warnCount);
              expect(getSuiteState(suiteId)).toMatchSnapshot();
            });
          });
        });
      });
    });
  });
});
