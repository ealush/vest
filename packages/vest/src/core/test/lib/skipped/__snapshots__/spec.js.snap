// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module: skipped export: mergeSkipped When previous state exists When skipped fields exist When skipped field exists in previous state Error field Should copy prev field state over 1`] = `
Object {
  "doneCallbacks": Array [],
  "errorCount": 1,
  "exclusive": Object {},
  "fieldCallbacks": Object {},
  "lagging": Array [],
  "name": "suite_1",
  "pending": Array [],
  "skipped": Object {
    "field_2": true,
  },
  "suiteId": "suiteId_1",
  "tests": Object {
    "field_2": Object {
      "errorCount": 1,
      "errors": Array [
        "error_string",
      ],
      "warnCount": 0,
      "warnings": Array [],
    },
  },
  "warnCount": 0,
}
`;

exports[`module: skipped export: mergeSkipped When previous state exists When skipped fields exist When skipped field exists in previous state Warning field Should copy prev field state over 1`] = `
Object {
  "doneCallbacks": Array [],
  "errorCount": 0,
  "exclusive": Object {},
  "fieldCallbacks": Object {},
  "lagging": Array [],
  "name": "suite_1",
  "pending": Array [],
  "skipped": Object {
    "field_1": true,
  },
  "suiteId": "suiteId_1",
  "tests": Object {
    "field_1": Object {
      "errorCount": 0,
      "errors": Array [],
      "warnCount": 1,
      "warnings": Array [
        "warning_string",
      ],
    },
  },
  "warnCount": 1,
}
`;

exports[`module: skipped export: setSkipped When field is not yet skipped Should marked field as skipped 1`] = `
Object {
  "doneCallbacks": Array [],
  "errorCount": 0,
  "exclusive": Object {},
  "fieldCallbacks": Object {},
  "lagging": Array [],
  "name": "suite_1",
  "pending": Array [],
  "skipped": Object {
    "field_1": true,
  },
  "suiteId": "suiteId_1",
  "tests": Object {},
  "warnCount": 0,
}
`;
